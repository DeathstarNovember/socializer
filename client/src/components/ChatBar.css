.chat-bar.sticky-top {
  top: 80px;
}

@media (max-device-width: 768px) {
  .chat-bar.sticky-top {
    height: 100%;
    padding-left: 15px;
    padding-right: 15px;
  }
}

@media (min-device-width: 768px) {
  .chat-bar.sticky-top {
    height: calc(100vh - 80px);
  }
}

.chat-conversations {
  overflow-y: scroll;
  border: 1px solid rgba(0, 0, 0, 0.125);
}

.chat,
.empty-state {
  border-top: 1px solid rgba(0, 0, 0, 0.125);
  border-bottom: 1px solid rgba(0, 0, 0, 0.125);
}

.chat:first-child {
  border-top: none;
}

.chat:last-child {
  border-bottom: none;
}

.chat:hover {
  text-decoration: none;
  color: #007bff;
  background-color: rgba(0, 0, 0, 0.05);
}

.chat:not(:last-child) {
  border-bottom: none;
}

.chat-avatars {
  flex-basis: 60px;
}

.chat-avatar-wrapper {
  width: 8px;
  overflow: visible;
}

.chat-title {
  flex: 1;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
