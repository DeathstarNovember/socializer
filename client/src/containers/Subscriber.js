import { useEffect } from "react";

const Subscriber = ({ subscribeToNew, children }) => {
  useEffect(() => {
    subscribeToNew();
  }, [subscribeToNew]);

  return children;
};

export default Subscriber;
