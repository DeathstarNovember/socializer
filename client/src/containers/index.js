export { default as StateProvider } from "./StateProvider";
export { default as Subscriber } from "./Subscriber";
