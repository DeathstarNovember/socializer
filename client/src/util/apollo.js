import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
