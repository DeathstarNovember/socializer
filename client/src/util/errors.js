import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
