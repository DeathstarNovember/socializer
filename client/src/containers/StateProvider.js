import React, { useEffect, useState } from "react";
import { useApolloClient } from "@apollo/react-hooks";
import Cookies from "js-cookie";
import { refreshSocket } from "util/apollo";
import { AuthContext, ChatContext } from "util/context";

const StateProvider = ({ initialToken, initialUserId, socket, children }) => {
  const client = useApolloClient();
  const [token, setToken] = useState(initialToken || Cookies.get("token"));
  const [userId, setUserId] = useState(initialUserId || Cookies.get("userId"));
  const [chatState, setChatState] = useState("default");

  // If the token changed (i.e. the user logged in
  // or out), clear the Apollo store and refresh the
  // websocket connection.
  useEffect(() => {
    if (!token) client.clearStore();
    if (socket) refreshSocket(socket);
  }, [client, socket, token]);

  const setAuth = (data) => {
    if (data) {
      const { id, token } = data;
      Cookies.set("token", token);
      Cookies.set("userId", id);
      setToken(token);
      setUserId(id);
    } else {
      Cookies.remove("token");
      Cookies.remove("userId");
      setToken(null);
      setUserId(null);
    }
  };

  return (
    <AuthContext.Provider value={{ token, userId, setAuth }}>
      <ChatContext.Provider value={{ chatState, setChatState }}>
        {children}
      </ChatContext.Provider>
    </AuthContext.Provider>
  );
};

export default StateProvider;
