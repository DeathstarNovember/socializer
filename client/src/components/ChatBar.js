import React, { useCallback, useContext } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Link } from "react-router-dom";
import { Button } from "react-bootstrap";
import Gravatar from "react-gravatar";
import produce from "immer";
import { reverse, sortBy } from "lodash";
import renderIf from "render-if";
import { NewConversation, QueryResult } from "components";
import { Subscriber } from "containers";
import { AuthContext, ChatContext } from "util/context";
import "./ChatBar.css";

export const GET_CONVERSATIONS = gql`
  query GetConversations {
    conversations {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

export const CONVERSATIONS_SUBSCRIPTION = gql`
  subscription onConversationCreated {
    conversationCreated {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

export const CONVERSATIONS_UPDATE_SUBSCRIPTION = gql`
  subscription onConversationUpdated {
    conversationUpdated {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

const ChatBar = () => {
  const { userId } = useContext(AuthContext);

  if (userId) {
    return <AuthChatBar />;
  } else {
    return <UnauthChatBar />;
  }
};

const UnauthChatBar = () => {
  return (
    <div className="p-2 empty-state">
      <h4>Chat</h4>
      <p>
        <Link to="/login">Log in</Link> to chat
      </p>
    </div>
  );
};

const AuthChatBar = () => {
  const { userId } = useContext(AuthContext);
  const { chatState, setChatState } = useContext(ChatContext);
  const { subscribeToMore, ...queryResult } = useQuery(GET_CONVERSATIONS);
  const subscribeToNew = useCallback(() => {
    subscribeToMore({
      document: CONVERSATIONS_SUBSCRIPTION,
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) return prev;
        const newConversation = subscriptionData.data.conversationCreated;

        // Check that we don't already have the
        // conversation stored.
        if (prev.conversations.find((c) => c.id === newConversation.id)) {
          return prev;
        }

        return produce(prev, (next) => {
          next.conversations.unshift(newConversation);
        });
      },
    });
    subscribeToMore({
      document: CONVERSATIONS_UPDATE_SUBSCRIPTION,
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) return prev;
        const updatedConversation = subscriptionData.data.conversationUpdated;
        return produce(prev, (next) => {
          const match = prev.conversations.findIndex(
            (c) => c.id === updatedConversation.id,
          );
          if (match !== -1) {
            next.conversations[match] = updatedConversation;
          }
        });
      },
    });
  }, []);

  return (
    <div className="chat-bar sticky-top d-flex flex-column pb-4">
      <h4 className="d-flex justify-content-between align-items-center">
        <span>{chatState === "creating" ? "New chat" : "Chat"}</span>
        {renderIf(chatState === "default")(
          <Button
            size="sm"
            variant="primary"
            onClick={() => setChatState("creating")}
          >
            New chat
          </Button>,
        )}
      </h4>

      <div className="chat-conversations d-flex flex-column">
        <QueryResult {...queryResult}>
          {({ data }) => (
            <Subscriber subscribeToNew={subscribeToNew}>
              {renderIf(chatState === "creating")(<NewConversation />)}

              {renderIf(chatState === "default")(
                reverse(sortBy(data.conversations, "updatedAt")).map(
                  (conversation) => (
                    <Link
                      key={conversation.id}
                      to={`/chat/${conversation.id}`}
                      className="d-flex align-items-center p-2 chat"
                    >
                      <div className="d-flex chat-avatars">
                        {conversation.users
                          .filter((user) => user.id !== userId)
                          .slice(0, 3)
                          .map((user) => (
                            <div key={user.id} className="chat-avatar-wrapper">
                              <Gravatar
                                md5={user.gravatarMd5}
                                className="rounded-circle"
                                size={30}
                              />
                            </div>
                          ))}
                      </div>
                      <div className="chat-title">{conversation.title}</div>
                    </Link>
                  ),
                ),
              )}
            </Subscriber>
          )}
        </QueryResult>
      </div>
    </div>
  );
};

export default ChatBar;
