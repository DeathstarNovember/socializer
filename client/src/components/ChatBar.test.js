import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { Subscriber } from "containers";
import { AuthContext, ChatContext } from "util/context";
import ChatBar, {
  GET_CONVERSATIONS,
  CONVERSATIONS_SUBSCRIPTION,
  CONVERSATIONS_UPDATE_SUBSCRIPTION,
} from "./ChatBar";

jest.mock("containers/Subscriber", () =>
  jest.fn().mockImplementation(({ children }) => children),
);

const getMock = {
  request: {
    query: GET_CONVERSATIONS,
  },
  result: {
    data: {
      conversations: [
        {
          id: 1,
          title: "Jane Smith, John Doe",
          updatedAt: "2019-04-18T00:00:00",
          users: [
            {
              id: 1,
              gravatarMd5: "abc",
            },
            {
              id: 2,
              gravatarMd5: "def",
            },
          ],
        },
      ],
    },
  },
};

const renderChatBar = ({ authContext, chatContext, mocks } = {}) => {
  return render(
    <MemoryRouter>
      <AuthContext.Provider value={authContext || { userId: 1 }}>
        <ChatContext.Provider
          value={
            chatContext || { chatState: "default", setChatState: jest.fn() }
          }
        >
          <MockedProvider mocks={mocks || [getMock]} addTypename={false}>
            <ChatBar />
          </MockedProvider>
        </ChatContext.Provider>
      </AuthContext.Provider>
    </MemoryRouter>,
  );
};

describe("ChatBar", () => {
  it("renders correctly when unauthenticated", () => {
    const { container } = renderChatBar({ authContext: {} });
    expect(container).toMatchSnapshot();
  });

  it("renders correctly for loading state", () => {
    const { container } = renderChatBar();
    expect(container).toMatchSnapshot();
  });

  it("renders correctly for loaded state", async () => {
    const { container, getByText } = renderChatBar();
    await wait(() => getByText("Jane Smith, John Doe"));
    expect(container).toMatchSnapshot();
  });

  it("renders correctly after created conversation", async () => {
    Subscriber.mockImplementation((props) => {
      const { default: ActualSubscriber } = jest.requireActual(
        "containers/Subscriber",
      );
      return <ActualSubscriber {...props} />;
    });

    const mocks = [
      getMock,
      {
        request: {
          query: CONVERSATIONS_SUBSCRIPTION,
        },
        result: {
          data: {
            conversationCreated: {
              id: 2,
              title: "Jane Smith, Jeremy Peters",
              updatedAt: "2019-04-19T00:00:00",
              users: [
                {
                  id: 1,
                  gravatarMd5: "abc",
                },
                {
                  id: 3,
                  gravatarMd5: "ghi",
                },
              ],
            },
          },
        },
      },
      {
        request: {
          query: CONVERSATIONS_UPDATE_SUBSCRIPTION,
        },
        result: {
          data: {
            conversationUpdated: {
              id: 1,
              title: "Jane Smith, John Doe",
              updatedAt: "2019-04-18T12:00:00",
              users: [
                {
                  id: 1,
                  gravatarMd5: "abc",
                },
                {
                  id: 2,
                  gravatarMd5: "def",
                },
              ],
            },
          },
        },
      },
    ];
    const { container, getByText } = renderChatBar({ mocks });
    await wait(() => getByText("Jane Smith, Jeremy Peters"));
    expect(container).toMatchSnapshot();
  });

  it("renders correctly for new chat state", () => {
    const { container } = renderChatBar({
      chatContext: {
        chatState: "creating",
        setChatState: jest.fn(),
      },
    });
    expect(container).toMatchSnapshot();
  });

  it("switches to new chat view", () => {
    const setChatState = jest.fn();
    const { getByText } = renderChatBar({
      chatContext: {
        chatState: "default",
        setChatState,
      },
    });
    fireEvent.click(getByText("New chat"));
    expect(setChatState).toBeCalledWith("creating");
  });
});
