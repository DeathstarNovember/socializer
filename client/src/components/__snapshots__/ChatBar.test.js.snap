// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBar renders correctly after created conversation 1`] = `
<div>
  <div
    class="chat-bar sticky-top d-flex flex-column pb-4"
  >
    <h4
      class="d-flex justify-content-between align-items-center"
    >
      <span>
        Chat
      </span>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        New chat
      </button>
    </h4>
    <div
      class="chat-conversations d-flex flex-column"
    >
      <a
        class="d-flex align-items-center p-2 chat"
        href="/chat/2"
      >
        <div
          class="d-flex chat-avatars"
        >
          <div
            class="chat-avatar-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle"
              height="30"
              src="//www.gravatar.com/avatar/ghi?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/ghi?d=retro&r=g&s=60 2x"
              width="30"
            />
          </div>
        </div>
        <div
          class="chat-title"
        >
          Jane Smith, Jeremy Peters
        </div>
      </a>
      <a
        class="d-flex align-items-center p-2 chat"
        href="/chat/1"
      >
        <div
          class="d-flex chat-avatars"
        >
          <div
            class="chat-avatar-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle"
              height="30"
              src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
              width="30"
            />
          </div>
        </div>
        <div
          class="chat-title"
        >
          Jane Smith, John Doe
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ChatBar renders correctly for loaded state 1`] = `
<div>
  <div
    class="chat-bar sticky-top d-flex flex-column pb-4"
  >
    <h4
      class="d-flex justify-content-between align-items-center"
    >
      <span>
        Chat
      </span>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        New chat
      </button>
    </h4>
    <div
      class="chat-conversations d-flex flex-column"
    >
      <a
        class="d-flex align-items-center p-2 chat"
        href="/chat/1"
      >
        <div
          class="d-flex chat-avatars"
        >
          <div
            class="chat-avatar-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle"
              height="30"
              src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
              width="30"
            />
          </div>
        </div>
        <div
          class="chat-title"
        >
          Jane Smith, John Doe
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ChatBar renders correctly for loading state 1`] = `
<div>
  <div
    class="chat-bar sticky-top d-flex flex-column pb-4"
  >
    <h4
      class="d-flex justify-content-between align-items-center"
    >
      <span>
        Chat
      </span>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        New chat
      </button>
    </h4>
    <div
      class="chat-conversations d-flex flex-column"
    >
      <div
        class="d-flex justify-content-center"
      >
        <div
          class="spinner-border"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatBar renders correctly for new chat state 1`] = `
<div>
  <div
    class="chat-bar sticky-top d-flex flex-column pb-4"
  >
    <h4
      class="d-flex justify-content-between align-items-center"
    >
      <span>
        New chat
      </span>
    </h4>
    <div
      class="chat-conversations d-flex flex-column"
    >
      <div
        class="d-flex justify-content-center"
      >
        <div
          class="spinner-border"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatBar renders correctly when unauthenticated 1`] = `
<div>
  <div
    class="p-2 empty-state"
  >
    <h4>
      Chat
    </h4>
    <p>
      <a
        href="/login"
      >
        Log in
      </a>
       to chat
    </p>
  </div>
</div>
`;
